core/src/main/java/org/apache/iceberg/BaseIncrementalAppendScan.java

```diff
@@ -0,0 +1,172 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.apache.iceberg.util.TableScanUtil;
+
+class BaseIncrementalAppendScan extends BaseScan<IncrementalAppendScan> implements IncrementalAppendScan {
+
+  BaseIncrementalAppendScan(TableOperations ops, Table table) {
+    this(ops, table, table.schema(), new TableScanContext());
+  }
+
+  BaseIncrementalAppendScan(TableOperations ops, Table table, Schema schema, TableScanContext context) {
+    super(ops, table, schema, context);
+  }
+
+  @Override
+  protected IncrementalAppendScan newRefinedScan(
+      TableOperations newOps, Table newTable, Schema newSchema, TableScanContext newContext) {
+    return new BaseIncrementalAppendScan(newOps, newTable, newSchema, newContext);
+  }
+
+  @Override
+  public IncrementalAppendScan fromSnapshotInclusive(long fromSnapshotId) {
+    Snapshot startingSnapshotInclusive = table().snapshot(fromSnapshotId);
+    Preconditions.checkArgument(startingSnapshotInclusive != null,
+        "Cannot find the starting snapshot: %d", fromSnapshotId);
+    // To achieve inclusive behavior, we pass the parent snapshot id to the TableScanContext.
+    // If parent id is null, leave the TableScanContext#fromSnapshotId as null.
+    // This way, TableScanContext#fromSnapshotId always capture the exclusive behavior.
+    if (startingSnapshotInclusive.parentId() != null) {
+      return newRefinedScan(tableOps(), table(), schema(),
+          context().fromSnapshotId(startingSnapshotInclusive.parentId()));
+    }
+    return this;
+  }
+
+  @Override
+  public IncrementalAppendScan fromSnapshotExclusive(long fromSnapshotId) {
+    Preconditions.checkArgument(table().snapshot(fromSnapshotId) != null,
+        "Cannot find starting snapshot: %d", fromSnapshotId);
+    return newRefinedScan(tableOps(), table(), schema(), context().fromSnapshotId(fromSnapshotId));
+  }
+
+  @Override
+  public IncrementalAppendScan toSnapshot(long toSnapshotId) {
+    Preconditions.checkArgument(table().snapshot(toSnapshotId) != null,
+        "Cannot find end snapshot: %d", toSnapshotId);
+    return newRefinedScan(tableOps(), table(), schema(),  context().toSnapshotId(toSnapshotId));
+  }
+
+  @Override
+  public CloseableIterable<FileScanTask> planFiles() {
+    // Return an empty iterable when table is empty
+    if (table().currentSnapshot() == null) {
+      return CloseableIterable.empty();
+    }
+
+    long toSnapshotIdInclusive = context().toSnapshotId() != null ?
+        context().toSnapshotId() : table().currentSnapshot().snapshotId();
+    if (context().fromSnapshotId() != null) {
+      Preconditions.checkArgument(
+          SnapshotUtil.isAncestorOf(table(), toSnapshotIdInclusive, context().fromSnapshotId()),
+          "Starting snapshot (exclusive) %d is not an ancestor of end snapshot %d",
+          context().fromSnapshotId(), toSnapshotIdInclusive);
+    }
```

Comment:
If the `fromSnapshotId` is null then we need to validate that the oldest ancestor's parent is null.

Reply:
Here's a case I'm thinking of:

```
 A - B - C <-- current
 D - E <-- branch
```

If D's parent is null and the `toSnapshotId` is C. If the from snapshot is D, then the exclusive one will be null, D's parent. But D is not in C's ancestors.

Reply:
Probably need to add an inclusive/exclusive flag to context so that we keep track of the requested starting snapshot, D.